node/network/statement-distribution/src/lib.rs

```diff
@@ -63,54 +82,177 @@ fn network_update_message(n: NetworkBridgeEvent) -> AllMessages {
 	AllMessages::StatementDistribution(StatementDistributionMessage::NetworkBridgeUpdate(n))
 }
 
+/// Tracks our impression of a single peer's view of the candidates a validator has seconded
+/// for a given relay-parent.
+///
+/// It is expected to receive at most `VC_THRESHOLD` from us and be aware of at most `VC_THRESHOLD`
+/// via other means.
+#[derive(Default)]
+struct VcPerPeerTracker {
+	local_observed: arrayvec::ArrayVec<[Hash; VC_THRESHOLD]>,
+	remote_observed: arrayvec::ArrayVec<[Hash; VC_THRESHOLD]>,
+}
+
+impl VcPerPeerTracker {
+	fn contains(&self, h: &Hash) -> bool {
+		self.local_observed.contains(h) || self.remote_observed.contains(h)
+	}
+
+	// Note that the remote should now be aware that a validator has seconded a given candidate (by hash)
+	// based on a message that we have sent it from our local pool.
+	fn note_local(&mut self, h: Hash) {
+		if self.local_observed.contains(&h) { return }
+
+		if self.local_observed.is_full() {
+			log::warn!("Statement distribution is erroneously attempting to distribute more \
+				than {} candidate(s) per validator index. Ignoring", VC_THRESHOLD);
+		} else {
+			self.local_observed.try_push(h).expect("length of storage guarded above; \
+				only panics if length exceeds capacity; qed");
+		}
+	}
+
+	// Note that the remote should now be aware that a validator has seconded a given candidate (by hash)
+	// based on a message that it has sent us.
+	//
+	// Returns `true` if the peer was allowed to send us such a message, `false` otherwise.
+	fn note_remote(&mut self, h: Hash) -> bool {
```

Comment:
Seems like we could factor the common elements of `note_local` and `note_remote` like

```rust
fn note_hash(container: &mut ArrayVec, h: Hash) -> bool {
  if container.contains(&h) { return true; }
  if container.is_full() { false }
  else { container.try_push(h).expect("guarded above"); true }
}

fn note_local(&mut self, h: Hash) {
  if ! Self::note_hash(&mut self.local_observed, h) {
    log::warn!("statement distribution error");
  }
}

fn note_remote(&mut self, h: Hash) -> bool {
  Self::note_hash(&mut self.remote_observed, h)
}
```

(not recommending the shortened log / expect strings; just didn't want to retype them all)